src/OrchardCore.Modules/OrchardCore.Media/Services/ChunkFileUploadBackgroundTask.cs

```diff
@@ -0,0 +1,18 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using OrchardCore.BackgroundTasks;
+
+namespace OrchardCore.Media.Services;
+
+[BackgroundTask(Schedule = "0/1 * * * *", Description = "Checks for abandoned file uploads.")]
```

Comment:
Maybe clean old temp files (with a timespan) on each new transfer then clean it. This way this task might not be necessary. (do it behind a lock so only one upload cleans at the same time)

Reply:
Purging on every upload can be slow for the user, and doesn't guard against people uploading a bunch of files at once, aborting them, and then not uploading anything new for a week (extreme, but still). So, I'd still keep the bg task.

Note to self: https://github.com/GovBuilt/GovBuilt-Platform/pull/309#discussion_r1262669051

Reply:
I think we can meet in the middle. I can add the bg task back to run once a day and also keep the current approach.

Reply:
Why is the bg task alone not enough though?

Reply:
Now the purge is invoked from only the bg task.

Reply:
OK then!

Reply:
But on upload the related file is still deleted, to avoid concurrency @dministro added a configurable `TemporaryFileLifetime` so that the purge doesn't collects "recent" files that may be still in use, and to prevent concurrent purges by multiple tenants the temp file names now include the `TenantId`.

For testing I tried to do a purge while uploading a file, no problem as long the file is in use it is just not deleted without exception, but here only one instance/process is running. Yes, this doesn't prevent file sharing violations if multiple running instances share the same file system, but there are try/catch blocks and errors logging.

The only thing I saw is that for a given image we always generate the same temp file name, so maybe a problem if we concurrently upload the same image, here maybe, as we differentiate per tenant, we could have differentiated per upload.

But it's a quite rare case, so all good for me.

Reply:
> The only thing I saw is that for a given image we always generate the same temp file name, so maybe a problem if we concurrently upload the same image, here maybe, as we differentiate per tenant, we could have differentiated per upload.

Did you reload the page before trying to upload the image again?

The temporary filename hash is the result of the `uploadId` (see below) the `fileName` and the file field name.

- `uploadId` - is a GUID and is generated on app loading/mounting. https://github.com/Lombiq/OrchardCore/blob/b54e4f4ff0d8cb2cc78ea5ffd49e60eec302f116/src/OrchardCore.Modules/OrchardCore.Media/Assets/js/MediaApp/fileupload.js#L2 and https://github.com/Lombiq/OrchardCore/blob/b54e4f4ff0d8cb2cc78ea5ffd49e60eec302f116/src/OrchardCore.Modules/OrchardCore.Media/Assets/js/MediaField/attachedMediaField.js#L132

@jtkech

Reply:
@dministro Okay cool then, indeed I didn't reload the page.